Better handling of an empty shard's segments_N file

abeyad opened the pull request:
When trying to restore a snapshot of an index created in a previous
version of Elasticsearch, it is possible that empty shards in the
snapshot have a segments_N file that has an unsupported Lucene version
and a missing checksum.  This leads to issues with restoring the
snapshot.  This commit handles this special case by avoiding a restore
of a shard that has no data, since there is nothing to restore anyway.

Closes #18707

abeyad commented:
@s1monw FYI

s1monw commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoveryFileChunkRequest.java:
```diff
@@ -105,11 +103,9 @@ public void readFrom(StreamInput in) throws IOException {
-        Version writtenBy = null;
-        String versionString = in.readOptionalString();
-        writtenBy = Lucene.parseVersionLenient(versionString, null);
+        Version writtenBy = Lucene.parseVersionLenient(in.readString(), null);
```
should we just use `Version.parseLenient(in.readString())` it must be non-null right?

abeyad replied:
Yes it has to be non-null, the only difference in using `Version.parseLeniently` being we will have to handle the `ParseException` ourselves.  I can catch the exception and wrap it into an `IOException` here

s1monw replied:
ok just add an assert here to be sure `assert writtenBy != null`

abeyad replied:
Done

s1monw commented on core/src/main/java/org/elasticsearch/index/store/StoreFileMetaData.java:
```diff
@@ -38,6 +36,8 @@
 
     public static final Version FIRST_LUCENE_CHECKSUM_VERSION = Version.LUCENE_5_0_0;
 
+    public static final String UNKNOWN_CHECKSUM = "_na_";
```
this should go away here I think it's impl details of a bwc layer in S/R

abeyad replied:
I'll add a method to `StoreFileMetaData` with something to the effect of `isUnknownChecksum`, then this field can be `private`.  Its still helpful to have a central place for it b/c other places use it (i.e. check on "_na_")

abeyad replied:
I see what you were saying now, i'll move this to `BlobStoreIndexShardSnapshot`, esp since I no longer need it for hashAndLengthEqual as its going away

s1monw commented on core/src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardSnapshot.java:
```diff
@@ -305,17 +314,37 @@ public static FileInfo fromXContent(XContentParser parser) throws IOException {
             } else if (length < 0) {
                 throw new ElasticsearchParseException("missing or invalid file length");
+            } else if (checksum == null) {
+                if (physicalName.startsWith("segments_")) {
```
can we verify `writtenBy` here too ie. if writtenBy is >= 6.0 we must fail since we are guaranteed to have the checksum here?

abeyad replied:
yes, good catch

s1monw commented on core/src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardSnapshot.java:
```diff
@@ -305,17 +314,37 @@ public static FileInfo fromXContent(XContentParser parser) throws IOException {
+                    // so we will assign it _na_ for now and try to get the checksum from the file itself later
+                    checksum = StoreFileMetaData.UNKNOWN_CHECKSUM;
+                } else {
+                    throw new ElasticsearchParseException("missing checksum for name [" + name + "]");
```
make this an IllegalStateException?

s1monw commented on core/src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -657,8 +654,8 @@ final void verifyAfterCleanup(MetadataSnapshot sourceMetaData, MetadataSnapshot
                     final boolean hashAndLengthEqual = (
-                            local.checksum() == null
-                                    && remote.checksum() == null
+                            local.checksum().equals(StoreFileMetaData.UNKNOWN_CHECKSUM)
```
why do we need this here? I think this entire `hashAndLengthEqual` can and has to go away

s1monw commented on core/src/test/java/org/elasticsearch/index/snapshots/blobstore/FileInfoTests.java:
```diff
@@ -105,11 +105,11 @@ public void testInvalidFieldsInFromXContent() throws IOException {
-            builder.field(Fields.LENGTH, length);
-            builder.field(Fields.WRITTEN_BY, Version.LATEST.toString());
-            builder.field(Fields.CHECKSUM, "666");
+            builder.field(BlobStoreIndexShardSnapshot.FileInfo.NAME, name);
```
why is this fully qualified?

s1monw commented:
left some comments - looks great!

abeyad commented:
@s1monw pushed up a new commit that addresses code review comments: https://github.com/elastic/elasticsearch/pull/18784/commits/f02e1ada947bde8eaa01ce5113407324a81f0146

Also, I reverted back to using `Lucene.parseVersionLenient` because using `Version.parseLeniently` is a forbidden API and `Lucene.parseVersionLenient` is the one central place its allowed.  Also, the `StoreFileMetaData` throws an NPE regardless in its constructor if `writtenBy` is `null`.

s1monw commented:
LGTM

abeyad commented:
@s1monw thanks for the review!

abeyad commented:
Closed by https://github.com/elastic/elasticsearch/commit/43e07c0c886223c65d26a4e3bdb7815ff15abe7d
